crates/red_knot/src/semantic/flow_graph.rs

```diff
@@ -109,43 +128,79 @@ impl FlowGraphBuilder {
     }
 }
 
+/// A definition, and the set of constraints between a use and the definition
+#[derive(Debug, Clone)]
+pub struct ConstrainedDefinition {
+    pub definition: Definition,
+    pub constraints: Vec<ExpressionId>,
+}
```

Comment:
Sure, I can play with this.

Reply:
I can easily remove the Vec from FlowState with this plan (and I'll push that), but I'm not sure about removing it from `ConstrainedDefinition`.

I think the part of this that can't work is borrowing the slice from the iterator; it seems like this is fundamentally not compatible with the design of the Iterator trait (see discussion at https://users.rust-lang.org/t/returning-borrowed-values-from-an-iterator/1096/10).

It might be possible to have `ConstrainedDefinition` also just hold a `Range<usize>`, and query the iterator for the actual constraints. This makes the API quite awkward (you have to make sure to do these queries before you iterate to the next definition, because of truncation), and I'm not sure it ends up any better in terms of allocations (because of the previous requirement, the caller would still end up just making its own Vec of the constraints anyway).

Do you think removing the vec from `ConstrainedDefinition` is important to solve now, or could be revisited post-Salsa?

Reply:
I think there's a reasonably good chance that post-Salsa, if we're doing inference all at once per scope, that the whole FlowGraph implementation may go away, in favor of eagerly tracking local symbol types as we walk the AST doing type inference. There's no reason to pay the overhead for laziness we don't use.

Reply:
Yeah you're right. You can't reference data stored on the vec itself. It either must have a longer lifetime or be per item. An alternative solution could be to change the item type to an `enum` where it is either a `Definition` or a `Constraint`, technically flattening the constraints. 

> Do you think removing the vec from ConstrainedDefinition is important to solve now, or could be revisited post-Salsa?

Not if you think it will go away with Salsa, maybe add a todo. Avoiding allocations for the CFG would be nice because it is a rather hot code path, at least in our current design.

Reply:
> An alternative solution could be to change the item type to an enum where it is either a Definition or a Constraint, technically flattening the constraints.

I think this could work, but I'm not sure how much it will reduce overall allocation. The caller will just end up having to collect the constraints as it waits for the definition they apply to, or create more intermediate intersection types along the way, which also involves allocation as intersection types hold a vec of their elements. 

It seems like we can shift the allocation around, but the basic need to have an arbitrary number of constraints apply to each definition means that we can't really get rid of it.